Core/TopBuilder.cpp

```diff
@@ -285,14 +285,17 @@ void Builder::accessChainStore(llvm::Value* value)
 {
     assert(accessChain.isRValue == false);
 
-    llvm::Value* base = collapseAccessChain();
-    llvm::Value* source = value;
-
-    // if swizzle exists, it is out-of-order or not full, we must load the target vector,
-    // extract and insert elements to perform writeMask and/or swizzle
-    if (accessChain.swizzle.size()) {
-
-        llvm::Value* shadowVector = createLoad(base, accessChain.metadataKind, accessChain.mdNode);
+    // if swizzle exists, it is out-of-order or not full, we must
+    // extract and store each element to perform writeMask and/or swizzle
+    if (accessChain.swizzle.size() > 1) {
+
+        llvm::Value* base = collapseAccessChain();
+        llvm::PointerType* outValTy = llvm::dyn_cast<llvm::PointerType>(base->getType());
+        assert(outValTy && "base of GEP should be pointer");
+        llvm::Type* vecType = base->getType()->getPointerElementType();
+        assert(llvm::isa<llvm::VectorType>(vecType));
+        llvm::Type* compPtrType = vecType->getVectorElementType()->getPointerTo(outValTy->getAddressSpace());
+        base = createBitCast(base, compPtrType);
```

Comment:
Everything should always be typed correctly.  Why do we need a bit cast?  Is a last (component selection) operand missing in creating the access chain?  If we don't have to start crossing the line from "all pointers working like a logical address" to "we're now doing real pointer stuff" in supporting shaders that aren't doing real pointer stuff, then it would be best to not start.